### PR TITLE
[VM] fix activatePrecompiles

### LINKED ISSUE #796: ActivatePrecompiles does not what it is expected to do
Assumption: if I create a VM with `activatePrecompiles` set to `true`, I expect that precompiles are "activated", which I expect to mean that the accounts are set to non-empty (if we call them from a contract, we expect that `callNewAccount` gas is **not** deducted). 

When we `activatePrecompiles`, every precompile address is filled with a `new Account()`. However:

```
let address = ("a").repeat(40)
vm.stateManager.putAccount(address, new Account())
vm.stateManager.accountIsEmpty(address).then(console.log)
```

=> `true`.

Hence the account is still empty.

Ganache fixes this by instantiating each account with a balance of nonzero. In that case the account is nonempty (we can also set the nonce of the account to 1, or set the codehash to something which is not `KECCAK256_NULL_S`, but these seem a bit weird). So I propose we instead put a balance of 1 wei in those accounts to properly activate them.

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Closes #796 

Can someone verify that `activatePrecompiles` is indeed what I expect it to mean? (See the issue)

### CODE REVIEW

File: packages/vm/lib/index.ts [hunk 1]
```diff
@@ -149,9 +149,18 @@ export default class VM extends AsyncEventEmitter {
 
     if (opts.activatePrecompiles && !opts.stateManager) {
       this.stateManager.checkpoint()
+      // put 1 wei in each of the precompiles in order to make the accounts non-empty and thus not have them deduct `callNewAccount` gas.
       Object.keys(precompiles)
         .map((k: string): Buffer => Buffer.from(k, 'hex'))
-        .forEach(async (k: Buffer) => await this.stateManager.putAccount(k, new Account()))
+        .forEach(
```
Comment (alcuadrado) [inline packages/vm/lib/index.ts:155]: This was already present, but this `forEach` introduces an unhandled promise for each precompile.

This should be something like:

```ts
     await Object.keys(precompiles)
      .map((k: string): Buffer => Buffer.from(k, 'hex'))
      .map(addr => this.stateManager.putAccount(
              addr,
              new Account({
                balance: '0x01',
              })
      ));
```

**Comment by alcuadrado** [comment]
As I mentioned in the issue, you are right about `activatePrecompiles`. I just left a comment about an already-existing issue.

It's really nice that you added a regression test!

File: packages/vm/lib/index.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/vm/lib/index.ts]: Good point! Changed! 😄

**Comment by holgerd77** [comment]
Have resolved the conflicts here and will merge once the tests pass, thanks Jochem, this looks really great!

One note: I was wondering if the tests rather belong in the constructor tests, since one could argue that they DO test the functionality from the constructor - hence - if the `activatePrecompile` init is working correctly. In this case one can nevertheless also argue to put them in `runCall.js` since it is testing the functionality itself from there. Just want to generally raise the case that we should look careful where we place tests and what is actually tested.

Some note beyond and just to get confirmation that I got this right: isn't this a super-major bug causing wrong gas calculation on EVERY run with activated precompiles where a precompile is used? Or do I miss something here? Then we should really give this some prominence in the next release notes.

Anyhow: great find Jochem! 😄

**Comment by alcuadrado** [comment]
It's not a huge consensus bug, but it can have a small impact for people using the vm to estimate gas or test gas contained transactions. Just slightly misleading results.

This flag is there to make the vm gas costs closest to mainnet by ensuring that the precompile accounts aren't empty and contacts don't pay more gas the first time a precompile is called.

**Comment by jochem-brouwer** [comment]
I had this issue either here or in some other PR, but it looks like it has to do with some timeout; all tests pass. When I restarted the CI it worked.

**Comment by jochem-brouwer** [comment]
@holgerd77 I don't think this is a consensus bug. We also have not seen anyone raise this point as an issue. I need to note that this is a super obscure "consensus" bug: it only makes sense if you call a precompile /with value/ (why would you do that?) and expect it to use 25000 gas less, because you expect the addresses to be "initialized" as you passed this option to the VM's constructor. It also does not fail on any of the state tests. (If you put 1 wei in these accounts, you change the genesis block).

I agree that it does not really have to do with testing runCall, the test is intended to test particularly if we activate the precompiles the right way. I don't really know how to group those tests though, but I agree you have a point here that it is not really a runCall test here.

**Pull request merged**